### ISSUE TITLE
Package information gets lost if method parameter is a static field

### ISSUE THREAD

**Issue by schakko** [INRIA/spoon#1257]
When passing a static field into a method, the package information of the accessed field is lost.

Static field:

	package my;

	public class StaticField {
		public final static String TEXT = "Test";
	}

# Working sample
Imported class "my.StaticField" is available in model and serialized:

	import my.StaticField;

	class TestWorking {
		public void literalConcat() {
			String r = "Hello:  " + StaticField.Text;
		}
	}

# Failing sample
Imported package "my" is missing in model:

	import my.StaticField;

	class TestNotWorking {
		private void method(String s) {
		}
		
		public void typeReference() {
			method(StaticField.TEXT);
		}
	}

## Expected result

	import my.StaticField;

	class TestNotWorking {
		private void method(String s) {
		}
		
		public void typeReference() {
			method(StaticField.TEXT);
		}
	}

## Actual result

	// <-- import missing

	class TestNotWorking {
		private void method(String s) {
		}
		
		public void typeReference() {
			method(StaticField.TEXT);
		}
	}

When serializing the class TestWorking, *StaticField* gets imported. This should be the behavior for the second class, too. But when class *TestNotWorking* is serialized, the statement "import my.StaticField" is missing.

**Comment by surli** [comment]
Hi @schakko,

Which version of Spoon did you use for your test? And which argument do you pass to Spoon: specifically, did you use no-classpath mode, and/or auto-import mode? 

And just to be sure, the class `TestNotWorking` is not in package `my`, right?

**Comment by schakko** [comment]
- I tried it with the latest snapshot 5.7.0 from Maven Repository and self-compiled the latest Github version. I also reproduced it with 5.6.0.
- I used --no-classpath for the execution
- Yes, *TestNotWorking* is not in the package "my" but in its parent package ("").

**Comment by surli** [comment]
Are you sure you don't use `--with-imports` args too?
Because without the arguments, I obtain the following codes:

```java
public class WorkingExample {
    public void literalConcat() {
        java.lang.String r = "Hello:  " + (spoon.test.imports.testclasses.other.my.StaticField.TEXT);
    }
}
```

```java
public class Example {
    private void method(java.lang.String s) {
    }

    public void typeReference() {
        method(spoon.test.imports.testclasses.other.my.StaticField.TEXT);
    }
}
```

And with imports, I obtain the right imports... You can have a look to the unit test I added.

**Comment by schakko** [comment]
Ah sorry, *--with-imports* is enabled. I'll try to take a look at the unit tests after work.

**Comment by surli** [comment]
Anything new here @schakko ?

**Comment by schakko** [comment]
Sorry for the late response, I just checked into the samples. First good news: the unit test does work when running in Eclipse or by using Maven. The output has the expected class definition. The test is correct.

The bad news is: I ran Spoon (not the tests) in Eclipse and on command line refereing to your unit test sample file by using

	-i <path to>\spoon\src\test\java\spoon\test\imports\testclasses\other\Example.java --noclasspath --with-imports -c

and the result is that the output is



	package spoon.test.imports.testclasses.other;

	import StaticField.TEXT;

	/**
	 * Created by urli on 10/04/2017.
	 */
	public class Example {
		private void method(String s) {
		}

		public void typeReference() {
			method(TEXT);
		}
	}

To sum it up: Unit test works as expected but real "test" does not work ;-)

**Comment by surli** [comment]
> `-i <path to>\spoon\src\test\java\spoon\test\imports\testclasses\other\Example.java --noclasspath --with-imports -c`

Actually you forget to tell Spoon to also use the StaticField java file as input: Spoon does not look for files but only use those which have been passed to it. 
So if you want to get the same behaviour, you need to specify as input all needed files:
```
-i ../src/test/java/spoon/test/imports/testclasses/other/Example.java:../src/test/java/spoon/test/imports/testclasses/other/my/StaticField.java --noclasspath --with-imports -c
```

or to be shorter, you can pass a directory to inspect:
```
-i ../src/test/java/spoon/test/imports/testclasses/other --noclasspath --with-imports -c
```

**Comment by schakko** [comment]
OK, but if I comment out the addInputResource in the unit test, the generated file does still contain the static field reference:

		final Launcher launcher = new Launcher();
		launcher.getEnvironment().setNoClasspath(true);
		launcher.getEnvironment().setAutoImports(true);
		String outputDir = "./target/spooned-staticnocp";
		// launcher.addInputResource("./src/test/java/spoon/test/imports/testclasses/other/my/StaticField.java");
		launcher.addInputResource("./src/test/java/spoon/test/imports/testclasses/other/Example.java");
		// launcher.addInputResource("./src/test/java/spoon/test/imports/testclasses/other/WorkingExample.java");
		launcher.setSourceOutputDirectory(outputDir);
		launcher.run();
		canBeBuilt(outputDir, 7);

produces

	package spoon.test.imports.testclasses.other;

	import spoon.test.imports.testclasses.other.my.StaticField;

	public class Example {
		private void method(String s) {
		}

		public void typeReference() {
			method(StaticField.TEXT);
		}
	}

**Comment by surli** [comment]
I just checked and you're right, but I think it's related to JDT and unit tests: those files are part of the code, then they are loaded when you launch unit testing, so JDT is able to retrieve the types because the class are loaded. 
However, when you launch Spoon with the JAR, test classes are not loaded then you need to specify the full list of input resources.

**Comment by surli** [comment]
I'm closing this one as it seems resolved to me, don't hesitate to reopen.
